esp32: Add support for board variants.

jimmo opened the pull request:
On the STM32 port, many of the pyboard board definitions support "variants" which enable extra features, such as threading, double-float, etc. The idea is to have a single board definition that can support different configurations.

This was recently added to the rp2 port, and used on the WEACTSTUDIO board for different flash sizes.

This PR adds this for the esp32 port, and merges the GENERIC boards down to just `GENERIC`, `GENERIC_S2`, `GENERIC_S3`, `GENERIC_C3`.

This has the following benefits:
 - There's just a single entry on the downloads page, rather than flooding the user with a whole bunch of nearly-indentical options. Then when selecting a variant the additional information in board.md can be shown to help choose.
 - It's easier to add variants than whole new board definitions. This is going to be important for the various combinations of flash and psram.

~TODO: Update tools/autobuild/build-boards.sh to query for variants and build them.~ Done. Removed the special case for stm32/pyboard and builds all variants for all boards across all ports.

_This work was funded through GitHub Sponsors._

jimmo committed:
esp32/boards/GENERIC_C3: Merge with GENERIC_C3_USB.

As the IDF no longer supports earlier revisions of the C3 by default, we
now just explicitly support rev 3+ and enable USB (which wasn't supported
in earlier revisions).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
esp32/boards/GENERIC: Merge with GENERIC_{SPIRAM,OTA,D2WD,UNICORE}.

These are now variants of the GENERIC board.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
esp32/boards/GENERIC_S2: Merge with ESP32_S2_WROVER.

Unsure of the history of the ESP32_S2_WROVER board (and why it wasn't
named GENERIC_S2_...) but now it's a variant of the generic S2 board.

Also removes the non-existent CONFIG_USB_AND_UART from all S2 boards.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
esp32/boards/GENERIC_S3: Merge with GENERIC_S3_{SPIRAM,SPIRAM_OCT}.

These are now variants of the GENERIC_S3 board.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/autobuild: Automatically build all variants for each board.

Removes the special-case for stm32.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Added the change to autobuild.sh to build all variants for all boards.

jimmo commented:
@dpgeorge This will require a small fixup on the build server to rename exisiting firmware files to the new names. But otherwise it should allow us to keep providing the full history for these "new" esp32 variants.

Flipje1955 commented:
Hi Jimmo, respectfully:  the statement "This was recently added to the rp2 port, and used on the WEACTSTUDIO board for different flash sizes." may have been implemented as an approach but it to date not followed through on, as the WEACTSTUDIO download page shows.

jimmo commented:
> Hi Jimmo, respectfully: the statement "This was recently added to the rp2 port, and used on the WEACTSTUDIO board for different flash sizes." may have been implemented as an approach but it to date not followed through on, as the WEACTSTUDIO download page shows.

hi @Flipje1955 -- this is exactly the purpose of the final commit in this PR. All variants for all boards in all ports will now be published to the downloads page. This includes the WEACTSTUDIO variants.

mattytrentini commented:
This is looking great @jimmo! It's a _very_ handy feature. 

However, don't we need additional variants defined for the GENERIC boards, in particular the S3? 4, 8, 16 and 32MB are all common configurations, I think...

jimmo commented:
> However, don't we need additional variants defined for the GENERIC boards, in particular the S3? 4, 8, 16 and 32MB are all common configurations, I think...

@mattytrentini Yep for sure. I just wanted to migrate the existing boards over as a first step.

jimmo committed:
esp32/Makefile: Implement `make submodules` to match other ports.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
all: Remove query-variants make target.

This is difficult to implement on cmake-based ports, and having the list
of variants in mpconfigboard.{cmake,mk} duplicates information that's
already in board.json.

This removes the existing query-variants make target from stm32 & rp2
and the definition of BOARD_VARIANTS from the various board files.

Also renames the cmake variable to MICROPY_BOARD_VARIANT to match other
variables such as MICROPY_BOARD. The make variable stays as
BOARD_VARIANT.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
After discussing with @dpgeorge we've decided to get rid of the `query-variants` target, and just use board.json to identify the variants that should be built. No point duplicating the list of variants in both `mpconfigboard.{mk,cmake}` as well as `board.json`.

So I've removed the definition of `BOARD_VARIANTS` from all boards that currently set it. I've kept the fix for `make submodules` for ESP32 in the PR.

Also rebased after #12081

mattytrentini commented:
First build I've tried appears to work:
```bash
> docker run -ti --rm -v $(pwd):/code -w /code espressif/idf:v5.0.2 bash -c "make -C mpy-cross && make -C ports/esp32 submodules all BOARD=GENERIC_S3 BOARD_VARIANT=spiram-oct"
...[snip]...
Project build complete. To flash, run this command:
/opt/esp/python_env/idf5.0_py3.8_env/bin/python ../../../opt/esp/idf/components/esptool_py/esptool/esptool.py -p (PORT) -b 460800 --before default_reset --after no_reset --chip esp32s3  write_flash --flash_mode dio --flash_size 8MB --flash_freq 80m 0x0 build-GENERIC_S3/bootloader/bootloader.bin 0x8000 build-GENERIC_S3/partition_table/partition-table.bin 0x10000 build-GENERIC_S3/micropython.bin
or run 'idf.py -p (PORT) flash'
bootloader  @0x000000    18288  (   14480 remaining)
partitions  @0x008000     3072  (    1024 remaining)
application @0x010000  1492224  (  539392 remaining)
total                  1557760
make: Leaving directory '/code/ports/esp32'
```

Looks _very_ promising!

jimmo commented:
> @dpgeorge This will require a small fixup on the build server to rename exisiting firmware files to the new names. But otherwise it should allow us to keep providing the full history for these "new" esp32 variants.

@dpgeorge For the `d2wd`, `ota`, `unicore` variants of GENERIC, no renaming necessary as the old boards had the exact same ID set in their `board.json` as the variant file will use. So just the `spiram` variant needs to be renamed from `esp32spiram`.

```bash
for f in esp32spiram-*.*; do echo mv $f ${f/esp32spiram/esp32-spiram}; done
```

Same for the `usb` variant of GENERIC_C3, nothing to be done.

For `GENERIC_S2`, the `ESP32_S2_WROVER` board is now the `wrover` variant:

```
for f in ESP32_S2_WROVER-*.*; do echo mv $f ${f/ESP32_S2_WROVER/GENERIC_S2-wrover}; done
```

For `GENERIC_S3`, the `GENERIC_S3_SPIRAM` and `GENERIC_S3_SPIRAM_OCT` boards are now the `spiram` and `spiram-oct` variants respectively:

```
for f in GENERIC_S3_SPIRAM_OCT-*.*; do echo mv $f ${f/GENERIC_S3_SPIRAM_OCT/GENERIC_S3-spiram-oct}; done
for f in GENERIC_S3_SPIRAM-*.*; do echo mv $f ${f/GENERIC_S3_SPIRAM/GENERIC_S3-spiram}; done
```

dpgeorge commented on extmod/extmod.cmake:
```diff
@@ -60,51 +60,62 @@ set(MICROPY_SOURCE_EXTMOD
-        ${MICROPY_LIB_BERKELEY_DIR}/btree/bt_utils.c
-        ${MICROPY_LIB_BERKELEY_DIR}/mpool/mpool.c
-    )
+    if(ECHO_SUBMODULES OR ECHO_QUERY_VARIANTS)
```
ECHO_QUERY_VARIANTS is not needed

dpgeorge commented on ports/esp32/boards/GENERIC/mpconfigboard.cmake:
```diff
@@ -2,3 +2,50 @@ set(SDKCONFIG_DEFAULTS
+endif()
+
+# Define the list of supported variants.
+set(BOARD_VARIANTS "d2wd ota spiram unicore")
```
not needed

dpgeorge commented on ports/esp32/boards/GENERIC/mpconfigboard.h:
```diff
@@ -1,2 +1,6 @@
+#define MICROPY_HW_BOARD_NAME "Generic ESP32 module"
+#endif
+
 #define MICROPY_HW_MCU_NAME "ESP32"
```
Should this be configured per-variant?  It previously did have different values for the variants, but arguably it should be the same for all of them, so probably OK to keep it at just a single value like you've done.

jimmo replied:
Yeah, so the d2wd and unicore boards did have different names before.

I think you could make the case that d2wd and unicore are attributes of the MCU whereas spiram and ota are attributes of the board/configuration. I think we should fix it.

In theory you could have e.g. a spiram+unicore variant, in which case I think setting the MCU to "esp32-unicore" and the board to "Generic ESP32 module with SPIRAM" makes sense.

Updated.

dpgeorge commented on ports/esp32/boards/GENERIC/mpconfigboard.cmake:
```diff
@@ -2,3 +2,50 @@ set(SDKCONFIG_DEFAULTS
+if(MICROPY_BOARD_VARIANT STREQUAL "d2wd")
+    set(SDKCONFIG_DEFAULTS
+        ${SDKCONFIG_DEFAULTS}
+        boards/GENERIC/sdkconfig.d2wd
```
this file is empty for some reason

dpgeorge commented on ports/esp32/boards/GENERIC_C3/mpconfigboard.cmake:
```diff
@@ -4,3 +4,13 @@ set(SDKCONFIG_DEFAULTS
+endif()
+
+# Define the list of supported variants.
+set(BOARD_VARIANTS "usb")
```
not needed

dpgeorge commented on ports/esp32/boards/GENERIC_C3/sdkconfig.c3usb:
```diff
@@ -1,4 +1,3 @@
-CONFIG_ESP32C3_REV_MIN_3=y
```
why is this line removed?

jimmo replied:
The simple answer is because it's the default anyway (without a bit of stuffing around you cannot compile firmware for revisions before 3).

I tried to piece the history together but I think what happened was that the engineering samples (rev 1 & 2) didn't support the USB peripheral, but our original GENERIC_C3 board definition was added then, and then later the GENERIC_C3_USB board was added to support the production boards.

Then later (IDF 5.0 I think, but haven't confirmed) the IDF made revision 3 the default minimum. i.e. neither GENERIC_C3 and GENERIC_C3_USB builds work by default any more on engineering samples. (I know this because I _only_ have engineering sample c3 boards).

So the longer answer is that I think we should actually switch this around and make usb the default (or possibly just not have a variant at all).

dpgeorge replied:
Hmm, OK.  So maybe we should follow IDF 5 and have the default GENERIC_C3 variant be minimum rev 3 and enable USB.  Then we could have a variant which runs on < rev 3 without USB.  Is that possible/easy?  If not easy then we can just have no variants for GENERIC_C3 and enable USB by default (until someone asks for < rev 3 support).

Said differently: if we only support rev 3, then we should enable USB by default.

jimmo replied:
Yes, I agree.

Will update later tonight.

jimmo replied:
Done.

dpgeorge commented on ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake:
```diff
@@ -4,3 +4,25 @@ set(SDKCONFIG_DEFAULTS
+endif()
+
+# Define the list of supported variants.
+set(BOARD_VARIANTS "wrover")
```
not needed

dpgeorge commented on ports/esp32/boards/GENERIC_S3/mpconfigboard.cmake:
```diff
@@ -6,3 +6,29 @@ set(SDKCONFIG_DEFAULTS
+endif()
+
+# Define the list of supported variants.
+set(BOARD_VARIANTS "spiram spiram-oct")
```
not needed

dpgeorge commented on py/mkrules.cmake:
```diff
@@ -178,7 +178,12 @@ if(MICROPY_FROZEN_MANIFEST)
     endif()
 
-    if(NOT (${ECHO_SUBMODULES}) AND NOT EXISTS ${MICROPY_LIB_DIR}/README.md)
+    if(ECHO_SUBMODULES OR ECHO_QUERY_VARIANTS)
```
ECHO_QUERY_VARIANTS not needed

dpgeorge commented on ports/esp32/boards/GENERIC_S2/mpconfigboard.cmake:
```diff
@@ -4,3 +4,25 @@ set(SDKCONFIG_DEFAULTS
+
+    list(APPEND MICROPY_DEF_BOARD
+        MICROPY_HW_I2C0_SCL=7
+        MICROPY_HW_I2C0_SDA=6
```
Did you consider putting settings like this in `mpconfigboard.h`?  Eg:
```c
#if MICROPY_BOARD_VARIANT_WROVER
#define MICROPY_HW_I2C0_SCL 7
#endif
```
Probably it's cleaner to keep all variant config in the .cmake file (and .mk for make-based ports)...

jimmo replied:
I did but there was no existing precedent for having the variant as a preprocessor macro, and exactly like you say it's neater putting it in one place.

jimmo replied:
I'm pretty confused by this "wrover" variant anyway and why it needs to define the pins?

dpgeorge replied:
Maybe because USB uses GPIO19 which is the default for I2C0 SDA?  But that would be a problem for all S2 boards.

jimmo replied:
How do you feel about replacing this variant with a "spiram" variant, which is consistent with what we've done for OG and S3?

jimmo replied:
I've pushed the branch with an extra commit that does this rename/replace. See the commit description for notes.

dpgeorge commented on ports/esp32/boards/GENERIC_S2/board.json:
```diff
@@ -14,5 +14,8 @@
+    "vendor": "Espressif",
+    "variants": {
+        "wrover": "WROVER module"
+    }
```
Looks like this change is in the wrong commit (it's in the GENERIC_S3 commit).

dpgeorge commented on ports/esp32/boards/GENERIC_S3_SPIRAM_OCT/mpconfigboard.cmake:
```diff
@@ -1,11 +0,0 @@
-    boards/sdkconfig.base
-    boards/sdkconfig.usb
-    boards/sdkconfig.ble
-    boards/sdkconfig.240mhz
```
this 240mhz file seems to be missing in the new variant

dpgeorge commented on tools/autobuild/build-boards.sh:
```diff
@@ -3,8 +3,36 @@
+#   $ cd ports/rp2
+#   $ MICROPY_AUTOBUILD_MAKE="make -j8" build_board boards/PICO/board.json -latest /tmp uf2
+
+function copy_artifacts {
```
The British spelling is "artefacts" :)

dpgeorge commented on tools/autobuild/build-boards.sh:
```diff
@@ -13,33 +41,29 @@ function build_board {
+
+    # Query variants from board.json and build them. Ignore the special "idf3"
+    # variant for ESP32 boards (this allows the downloads page to still the
+    # idf3 files for older releases that used to be explicitly built).
```
minor grammatical error, "to still have the"

jimmo commented:
Updated with review comments fixed.

jimmo committed:
esp32/partitions.csv: Rename to partitions-4MiB.csv.

To be consistent with the other partitions files (which have the "-
{2,8,16,32}MiB" suffix). Also renames partitions-ota.csv.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Updated based on further discussion.

For everything except OG, there is no longer a "spiram" variant. For simplicity, we just support spiram on the default configuration and it auto-detects at boot.      (The OG needs to maintain a spiram variant because enabling spiram comes at additional costs to work around silicon bugs that you don't want to pay for if you don't have spiram).

Added a board.md for all the generic boards that explains the different variants and how they map to the Espressif module names.

Squashed the S2_WROVER commit into the main S2 commit.

Renamed the default `partitions.csv` to `partitions-4MiB.csv` to match the others.

dpgeorge commented on ports/esp32/boards/GENERIC_S2/mpconfigboard.h:
```diff
@@ -1,8 +1,5 @@
-#define MICROPY_HW_ENABLE_SDCARD            (0)
-
-// Enable UART REPL for modules that have an external USB-UART and don't use native USB.
-#define MICROPY_HW_ENABLE_UART_REPL         (1)
```
I see you removed this... what was the reason?

Related to this, it seems that `CONFIG_USB_AND_UART` is not a real thing, at least I could not find it in the IDF or TinyUSB component...

jimmo replied:
Yep, that should be still there. Updated

Also removed `CONFIG_USB_AND_UART` from all S2 boards.

dpgeorge commented:
Merged, thanks for all the work on this one!

robert-hh commented:
@jimmo I have a question about using this new board variant scheme. If there a port has a basic configuration and variants, does the basic configuration has to be referred in boards.json as well?

jimmo commented:
@robert-hh Nope it doesn't.
